### PR TITLE
cmake: rewrite Findgenl to support components argument

### PR DESCRIPTION
**Pull request by tchaikov**
* rename genl to nl:
  "genl" is not very specific without more context. and libnl is actually
  a library suite. so it would be better if we can rename the library to
  be found to "libnl", and by following the naming convention of CMake,
  in this change, Findgenl.cmake is renamed to Findnl.cmake
* Findnl.cmake: support the component argument, in our specific case,
  what we want is "libnl-genl" library, which is one of the libraries
  offered by libnl. so let's just make it a component.
* Findnl.cmake: should pass the names of the required variables instead of
  their values to `find_package_handle_standard_args()`. before this
  change, we pass `GENL_LIBRARIES` to this function. it was correct at the
  first glance. but it is not able to handle the case where case where
  libnl-genl is not installed. so the fix is to pass all the names of
  required library paths to this function. in this change, their name
  are concatenated to a single variable -- `nl_LIBRARIES`, and the
  value of this variable is passed to
  `find_package_handle_standard_args()`. and the error message would
  be more specific this way if libnl-genl is not found:
    Could NOT find nl (missing: nl_genl_LIBRARY)
* Findnl.cmake: add nl::<component> as imported library, it helps the
  consumer of these libraries to import them -- no need to
  specify the `target_include_directories()` separately anymore.
* move the find_package() code to where it is used. it helps to improve
  the readability.
* remove `HAVE_GENL` variable: it's not used anywhere.
* drop the messages of "Not using foobar", they do not help.

Signed-off-by: Kefu Chai <kchai@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
hey @mikechristie could you help take a look?

File: debian/control [hunk 1]
```diff
@@ -53,7 +53,6 @@ Build-Depends: cmake (>= 3.5),
                libssl-dev,
                libtool,
                libudev-dev,
-               libnl-3-dev,
```
Comment (mikechristie) [inline debian/control]: I didn't get this part. Isn't it needed so on debian/ubuntu that package gets brought in automatically like it is for rpm based systems?

File: cmake/modules/Findnl.cmake [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    libnl3)
+
+# find the core library by default
+list(FIND nl_FIND_COMPONENTS "core" found)
+if(found EQUAL -1)
+  list(APPEND nl_FIND_COMPONENTS "core")
+endif()
+
+foreach(component ${nl_FIND_COMPONENTS})
+  set(nl_COMPONENTS core cli genl idiag nf route xfrm)
```
Comment (mikechristie) [inline cmake/modules/Findnl.cmake]: Hey, I am not very familiar with cmake. That's why the original code looked very basic like the udev/blkid cmake stuff :)

For rbd-nbd we only use genl. Is it necessary to search for these other components or is more a cmake standard type of style?

File: debian/control (hunk 1 above)
Comment (mikechristie) [reply debian/control]: Ignore this comment. I see it in the git commit.

File: cmake/modules/Findnl.cmake (hunk 2 above)
Comment (mikechristie) [reply cmake/modules/Findnl.cmake]: I meant to say we only use libnl and libnl-genl.

Comment (tchaikov) [reply cmake/modules/Findnl.cmake]: > Is it necessary to search for these other components

no, it's not necessary. they are listed here only for the sake of completeness.

> we only use libnl and libnl-genl.

yeah, i agree. i would be happy to remove the others if you think it makes the script easier to understand or better.

Comment (mikechristie) [reply cmake/modules/Findnl.cmake]: It does not matter to me.

**Comment by tchaikov** [comment]
```
make -f src/tools/rbd_nbd/CMakeFiles/rbd-nbd.dir/build.make src/tools/rbd_nbd/CMakeFiles/rbd-nbd.dir/depend
make[4]: Entering directory '/build/ceph-15.0.0-1803-ge46816e/obj-x86_64-linux-gnu'
src/tools/rbd_nbd/CMakeFiles/rbd-nbd.dir/build.make:99: *** target pattern contains no '%'.  Stop.
```

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=huge/27075//consoleFull

**Comment by tchaikov** [comment]
@dillaman ping?

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-cmake-detect-libnl/a17ba69e15b5da3ea75e7937ae38c43fa515e846/

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-cmake-detect-libnl/86b62af5dea9688d010d014b4eea77dacd14e703/

**Comment by tchaikov** [comment]
retest this please.

**Pull request merged**